src/dbnode/storage/namespace.go

```diff
@@ -1602,3 +1604,171 @@ func (n *dbNamespace) FlushState(shardID uint32, blockStart time.Time) (fileOpSt
+	ctx context.Context,
+	sourceNs databaseNamespace,
+	opts AggregateTilesOptions,
+	pm persist.Manager,
+) (int64, error) {
+	callStart := n.nowFn()
+	processedBlockCount, err := n.aggregateTiles(ctx, sourceNs, opts, pm)
+	n.metrics.writeAggData.ReportSuccessOrError(err, n.nowFn().Sub(callStart))
+
+	return processedBlockCount, err
+}
+
+func (n *dbNamespace) aggregateTiles(
+	ctx context.Context,
+	sourceNs databaseNamespace,
+	opts AggregateTilesOptions,
+	pm persist.Manager,
+) (int64, error) {
+	targetBlockSize := n.Metadata().Options().RetentionOptions().BlockSize()
+	blockStart := opts.Start.Truncate(targetBlockSize)
+	if blockStart.Add(targetBlockSize).Before(opts.End) {
+		return 0, fmt.Errorf("tile aggregation must be done within a single target block (start=%s, end=%s, blockSize=%s)",
+			opts.Start, opts.End, targetBlockSize.String())
+	}
+
+	n.RLock()
+	if n.bootstrapState != Bootstrapped {
+		n.RUnlock()
+		return 0, errNamespaceNotBootstrapped
+	}
+	nsCtx := n.nsContextWithRLock()
+	n.RUnlock()
+
+	targetShards := n.OwnedShards()
+
+	// Note: Cold writes must be enabled for Large Tiles to work.
+	if !n.nopts.ColdWritesEnabled() {
+		return 0, errColdWritesDisabled
+	}
+
+	sourceBlockSize := sourceNs.Metadata().Options().RetentionOptions().BlockSize()
+	sourceBlockStart := opts.Start.Truncate(sourceBlockSize)
+
+	sourceNsOpts := sourceNs.StorageOptions()
+	reader, err := fs.NewReader(sourceNsOpts.BytesPool(), sourceNsOpts.CommitLogOptions().FilesystemOptions())
+	if err != nil {
+		return 0, err
+	}
+
+	wOpts := series.WriteOptions{
+		TruncateType: n.opts.TruncateType(),
+		SchemaDesc:   nsCtx.Schema,
+	}
+
+	resources, err := newColdFlushReuseableResources(n.opts)
+	if err != nil {
+		return 0, err
+	}
+
+	// NB(bodu): Deferred targetShard cold flushes so that we can ensure that cold flush index data is
+	// persisted before persisting TSDB data to ensure crash consistency.
+	multiErr := xerrors.NewMultiError()
+	var processedBlockCount int64
+	for _, targetShard := range targetShards {
+		sourceShard, _, err := sourceNs.readableShardAt(targetShard.ID())
+		if err != nil {
+			detailedErr := fmt.Errorf("no matching shard in source namespace %s: %v", sourceNs.ID(), err)
+			multiErr = multiErr.Add(detailedErr)
+			continue
+		}
+		shardProcessedBlockCount, err := targetShard.AggregateTiles(ctx, reader, sourceNs.ID(), sourceBlockStart, sourceShard, opts, wOpts)
+		processedBlockCount += shardProcessedBlockCount
+		if err != nil {
+			detailedErr := fmt.Errorf("shard %d aggregation failed: %v", targetShard.ID(), err)
+			multiErr = multiErr.Add(detailedErr)
+			continue
+		}
+
+		multiErr = n.coldFlushSingleShard(nsCtx, targetShard, pm, resources, multiErr)
+	}
+
+	return processedBlockCount, multiErr.FinalError()
+}
+
+func (n *dbNamespace) coldFlushSingleShard(
+	nsCtx namespace.Context,
+	shard databaseShard,
+	pm persist.Manager,
+	resources coldFlushReuseableResources,
+	multiErr xerrors.MultiError,
```

Comment:
nit: instead of passing around MultiErrors, it may be a bit cleaner if `coldFlushSingleShard` just return an error on failure and then `aggregateTiles` can collect them into a multierror itself?

Reply:
We are no longer using cold flushes in the latest versions of aggregation (@gediminasgu could you confirm this?), so I guess there is no point in fixing the code that is being deleted.

Reply:
That's right. If possible, let's ignore this part of code as later it will replaced with a large tiles writer.